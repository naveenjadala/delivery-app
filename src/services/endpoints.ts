interface Endpoints {
  [key: string]: string;
}

const endpoints: Endpoints = {
  loginApi: '/user/login',
  SIGNUP: '/user/register',
  GET_ALL_USERS: '/user/getUsers',
  GET_RESTAURANTS: '/restaurants',
  SAVE_RESTAURANTS : '/restaurants/restaurants',
};

export default endpoints;
