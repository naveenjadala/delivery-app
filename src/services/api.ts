import axios from 'axios';
import endpoints from './endpoints';
import {retrieveData} from '../storage';

const API_URL = 'http://localhost:8000/api/';

const apiClient = axios.create({
  baseURL: API_URL,
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json',
  },
});

apiClient.interceptors.request.use(
  async function (config) {
    if (config.url !== '/user/login') {
      const token = await retrieveData('token');
      if (token) {
        config.headers['Authorization'] = `Bearer ${token}`;
      }
    }
    return config;
  },
  function (error) {
    return Promise.reject(error);
  },
);

interface loginData {
  email: string;
  password: string;
}

interface restaurants {
  name: string;
  location: string;
}

const login = async (endpoint: string, data: string): Promise<any> => {
  try {
    const response = await apiClient.post(`${endpoints[endpoint]}`, data);
    return response.data;
  } catch (error: any) {
    throw new Error(`API request to ${endpoint} failed${error.message}`);
  }
};

const signUp = async (endpoint: string, data: loginData): Promise<any> => {
  try {
    const response = await apiClient.post(`/${endpoints[endpoint]}`, data);
    return response;
  } catch (error) {
    throw new Error('API request to ${endpoint} failed${error.message}');
  }
};

const getAllUsers = async (endpoint: string) => {
  try {
    const response = await apiClient.get(`${endpoints[endpoint]}`);
    return response.data;
  } catch (error: any) {
    throw new Error(`API request for ${endpoint} failed with ${error.message}`);
  }
};

const getRestaurants = async (endpoint: string) => {
  try {
    const response = await apiClient.get(`${endpoints[endpoint]}`);
    return response.data.restaurants;
  } catch (error: any) {
    console.log(`API request for ${endpoint} failed with ${error}`);
    // throw new Error(`API request for ${endpoint} failed with ${error}`);
  }
};

const saveRestaurants = async (endpoint: string, data: restaurants) => {
  try {
    const response = await apiClient.post(`${endpoints[endpoint]}`, data);
    return response.data;
  } catch (error: any) {
    throw new Error(`API request to ${endpoint} failed${error.message}`);
  }
};

export {login, signUp, getAllUsers, getRestaurants, saveRestaurants};
