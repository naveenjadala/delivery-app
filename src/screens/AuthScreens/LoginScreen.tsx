import React, { useContext, useEffect, useRef, useState } from 'react'
import { Animated, Button, KeyboardAvoidingView, Platform, SafeAreaView, ScrollView, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native'
import { login } from '../../services/api';
import { Formik } from 'formik';
import { primaryColor } from '../../assets/color';
import { dynamicFontSize } from '../../utils/DynamicStylingUtils';
import { AuthContext } from './AuthContext';
import SubmitBtn from '../../components/Buttons/SubmitBtn';
import CustomeFormikTextInput from '../../components/TextInputs/CustomeFormikTextInput';

type loginProps = {
    email: string,
    password: string
}

const LoginScreen = () => {
    const ScaleValue = useRef(new Animated.Value(0)).current;
    const { setToken } = useContext(AuthContext);

    const loginApiCall = async (values: loginProps) => {
        try {
            let data = JSON.stringify({
                email: values.email,
                password: values.password
            });
            const logindata = await login('loginApi', data);
            if (logindata.status === 'success') {
                setToken(logindata.token);
            }
        } catch (error: any) {
            console.error("Error will login", error);
        }
    }

    const startAnimation = () => {
        Animated.timing(ScaleValue, {
            toValue: 1,
            duration: 3000,
            useNativeDriver: true,
        }).start();
    }

    useEffect(() => {
        startAnimation();
    }, []);

    return (
        <SafeAreaView style={{ flex: 1 }}>
            <ScrollView
                contentContainerStyle={{ flexGrow: 1 }}
                keyboardShouldPersistTaps="handled">
                <KeyboardAvoidingView
                    behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
                    style={styles.container}>
                    <Formik
                        initialValues={{
                            email: '',
                            password: '',
                        }}
                        onSubmit={values => loginApiCall(values)}>
                        {({ handleChange, handleSubmit, values }) => (

                            <View style={styles.formContainer}>
                                <CustomeFormikTextInput
                                    placeholder='Email Address'
                                    keyboardType='email-address'
                                    onChangeText={handleChange('email')}
                                    value={values.email} />
                                    
                                <CustomeFormikTextInput
                                    secureTextEntry
                                    placeholder='Password'
                                    onChangeText={handleChange('password')}
                                    value={values.password} />

                                <Animated.View style={{ marginVertical: 20, opacity: ScaleValue }}>
                                    <SubmitBtn lable="Login" onPress={() => handleSubmit()} />
                                </Animated.View>
                            </View>
                        )}
                    </Formik>
                </KeyboardAvoidingView>
            </ScrollView>
        </SafeAreaView>
    )
}

export default LoginScreen

const styles = StyleSheet.create({
    container: {
        flex: 1,
        // justifyContent: 'center',
        alignItems: 'center',
    },
    formContainer: {
        marginTop: 20,
        width: '80%',
        // alignItems: 'center',
        // alignSelf: 'center',
    },
})