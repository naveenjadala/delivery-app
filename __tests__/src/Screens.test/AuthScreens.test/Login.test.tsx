import React from 'react';
import {
  render,
  fireEvent,
  waitFor,
  act,
  screen,
} from '@testing-library/react-native';
import LoginScreen from '../../../../src/screens/AuthScreens/LoginScreen';
import * as api from '../../../../src/services/api';

jest.useFakeTimers();

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => {
    return jest.fn;
  },
}));

jest.mock('../../../../src/services/api', () => {
  const login = jest.fn(() =>
    Promise.resolve({ status: 'success', token: 'mockToken' }),
  );
  return {
    login,
  };
});

describe('<LoginScreen />', () => {
  //snapshot
  it('should render correctly', () => {
    const { toJSON } = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  //intigration test
  it('should allow user input in the email and password field', async () => {
    render(<LoginScreen />);

    const emailInput = await screen.findByPlaceholderText('Email Address');
    const passwordInput = await screen.findByPlaceholderText('Password');

    await act(async () => {
      fireEvent.changeText(emailInput, 'test@gmail.com');
      fireEvent.changeText(passwordInput, 'password123');
    });
    expect(emailInput.props.value).toBe('test@gmail.com');
    expect(passwordInput.props.value).toBe('password123');
  });

  // service call
  it('Form submission and service call', async () => {
    const { getByPlaceholderText, getByText } = render(<LoginScreen />);
    const emailInput = getByPlaceholderText('Email Address');
    const passwordInput = getByPlaceholderText('Password');

    fireEvent.changeText(emailInput, 'jadala@gmail.com');
    fireEvent.changeText(passwordInput, 'naveen');
    fireEvent.press(getByText('Login'));

    await waitFor(() =>
      expect(api.login).toHaveBeenCalledWith(
        'loginApi',
        JSON.stringify({ email: 'jadala@gmail.com', password: 'naveen' }),
      ),
    );
  });

  // afterEach(() => {
  //     jest.runOnlyPendingTimers();
  // });
});
