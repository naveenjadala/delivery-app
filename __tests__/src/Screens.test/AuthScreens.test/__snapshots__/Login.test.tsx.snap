// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should render correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
            },
            {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "marginTop": 20,
              "width": "80%",
            }
          }
        >
          <TextInput
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="Email Address"
            style={
              {
                "borderBottomWidth": 0.5,
                "borderColor": "#000",
                "marginVertical": 15,
              }
            }
            value=""
          />
          <TextInput
            onChangeText={[Function]}
            placeholder="Password"
            secureTextEntry={true}
            style={
              {
                "borderBottomWidth": 0.5,
                "borderColor": "#000",
                "marginVertical": 15,
              }
            }
            value=""
          />
          <View
            collapsable={false}
            style={
              {
                "marginVertical": 20,
                "opacity": 0,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FA4A0C",
                  "borderRadius": 15,
                  "height": 50,
                  "justifyContent": "center",
                  "marginVertical": 20,
                  "opacity": 1,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontSize": 33,
                  }
                }
              >
                Login
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
